### ISSUE TITLE
Strange type inference bug with unions and ifs

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#7414]
**Describe the bug**

<details>
<summary>Example code</summary>

```python
from __future__ import annotations

from typing import Union
from typing_extensions import Literal, reveal_type


class Invoke:
    ...


class StepDelta:
    ...


class MessageDeltaEvent:
    ...


class InvokeStepDeltaEvent:
    delta: StepDelta


class InvokeCompleted:
    data: Invoke

    event: Literal["invoke.completed"]


class MessageDelta:
    data: MessageDeltaEvent

    event: Literal["message.delta"]


class InvokeStepDelta:
    data: InvokeStepDeltaEvent

    event: Literal["invoke.step.delta"]


Event = Union[
    InvokeCompleted,
    InvokeStepDelta,
    MessageDelta,
]


class Foo:
    def handler(self, event: Event) -> None:
        if event.event == "invoke.completed":
            pass

        if event.event == "message.delta":
            if event.data:
                ...
        elif event.event == "invoke.step.delta":
            reveal_type(event.data.delta)
```

</details>

Pyright reports these errors:

```
foo.py
  foo.py:57:36 - error: Cannot access member "delta" for type "MessageDeltaEvent"
    Member "delta" is unknown (reportGeneralTypeIssues)
  foo.py:57:25 - error: Type of "delta" is unknown (reportUnknownMemberType)
  foo.py:57:25 - information: Type of "event.data.delta" is "Unknown"
2 errors, 0 warnings, 1 information
```

So for some reason it thinks that the `event.data` is of type `MessageDeltaEvent` when it should be `InvokeStepDeltaEvent`.

Strangely, removing the `if event.data` or the `if event.event == "invoke.completed"` blocks makes the error go away.

**VS Code extension or command-line**

From bisecting, this first started reporting an error in `v1.1.274`, earlier versions will output the correct type
```
foo.py
  foo.py:57:25 - information: Type of "event.data.delta" is "StepDelta"
0 errors, 0 warnings, 1 information
```

**Comment by erictraut** [comment]
Thanks for the bug report. This is a strange one indeed. I've traced its cause to a bug in pyright's code flow engine. The fix is going to be a bit tricky, so it may take me some time.

**Comment by erictraut** [comment]
This is included in pyright 1.1.353, which I just published.
